packages/nimble-components/src/rich-text/viewer/index.ts

```diff
@@ -39,15 +60,93 @@ export class RichTextViewer extends FoundationElement {
      * @internal
      */
     public markdownChanged(): void {
-        if (this.$fastController.isConnected) {
+        this.updateView();
+    }
+
+    /**
+     * @internal
+     */
+    public handleChange(source: unknown, args: unknown): void {
+        if (source instanceof MentionInternals && typeof args === 'string') {
+            this.updateMentionInternalsConfig();
+        }
+    }
+
+    private childItemsChanged(): void {
+        void this.updateMentionsFromChildItems();
+    }
+
+    private async updateMentionsFromChildItems(): Promise<void> {
+        const definedElements = this.childItems.map(async item => (item.matches(':not(:defined)')
+            ? customElements.whenDefined(item.localName)
+            : Promise.resolve()));
+        await Promise.all(definedElements);
+        this.mentionElements = this.childItems.filter(
+            (x): x is RichTextMention => x instanceof RichTextMention
+        );
+
+        this.observeMentions();
+        this.updateMentionInternalsConfig();
+    }
+
+    private observeMentions(): void {
+        this.removeMentionObservers();
+
+        for (const mention of this.mentionElements) {
+            const notifierInternals = Observable.getNotifier(
+                mention.mentionInternals
+            );
+            notifierInternals.subscribe(this);
+            this.mentionInternalsNotifiers.push(notifierInternals);
+        }
+    }
+
+    private removeMentionObservers(): void {
+        this.mentionInternalsNotifiers.forEach(notifier => {
+            notifier.unsubscribe(this);
+        });
+        this.mentionInternalsNotifiers = [];
+    }
+
+    private updateMentionInternalsConfig(): void {
+        this.mentionInternalsConfig = [];
+
+        if (this.hasDuplicateConfigurationElement()) {
             this.updateView();
+            return;
         }
+
+        this.mentionInternalsConfig = this.mentionElements
+            .filter(mention => mention.mentionInternals.validConfiguration)
+            .map(
+                mention => new MarkdownParserMentionConfiguration(
+                    mention.mentionInternals
+                )
+            );
+
+        this.updateView();
+    }
+
+    private hasDuplicateConfigurationElement(): boolean {
```

Comment:
It is because all the validation handled in the mention-validator holds specific to the configuration element and mapping elements, not on the number of configuration elements within the children of the viewer component.

The problem we see in not including it in the mention-validator is that the creation of mention configs happens only based on the valid state of the elements exactly at [this part of the code](https://github.com/ni/nimble/blob/d5330e74cfc855856bed26fc6b6f2fe3dde2eadd/packages/nimble-components/src/rich-text-mention/base/index.ts#L106). Also, we can identify if there are duplicate configuration elements only in the parent component (viewer/editor) so we are not sure how can utilize the mention-validator for this case because each configuration has its own validation that might not be able to determine what element in the children is duplicate.

We are validating this at the top layer (`childItemsChanged`) if there are duplicate configurations found and avoid other processes like observing the mention internals change or updating the mention configurations. Only after the duplicates are removed from the children, the subscriptions are added, then listen to the mention internals state.

Reply:
Gotcha, that makes sense that an individual mention configuration can't validate this state because it depends on its siblings. But I think the right place to validate it is in a validator for the rich text viewer. This is preferable to your current approach because it allows clients to inspect the validator to see why their mentions aren't rendering.

This pattern is similar to how [the table has a validator](https://github.com/ni/nimble/blob/main/packages/nimble-components/src/table/models/table-validator.ts) which handles table validity and aggregates validity of its columns (see `invalidColumnConfiguration`). But now the rich text viewer will have a validator which handles this case and also aggregates the validity of the mention config elements.

Doing that would be a good sized PR on its own. Ideally it would be a separate PR that precedes this one. But I think it'd also be acceptable to remove the code from this PR, leave a TODO comment describing this case, and do something sane for now like pick the first config element that matches the mention character. Then in a follow up PR you could implement the validation I described above and remove the TODO.

Reply:
It makes sense to have it as a validator for the viewer component like the table validator. Thanks for sharing the references. 

As suggested, I have removed the validation from here, added a TODO in the files needed, and will handle this validation in the follow-up PR.